### PR TITLE
Extend TagForReplaceMode to adapt Databricks runtime 

### LINKED ISSUE #3339: [BUG] Databricks test fails test_hash_groupby_collect_partial_replace_fallback
10:05:47  FAILED ../../src/main/python/hash_aggregate_test.py::test_hash_groupby_collect_partial_replace_fallback[false-false-{'spark.rapids.sql.variableFloatAgg.enabled': 'true', 'spark.rapids.sql.hasNans': 'true', 'spark.rapids.sql.castStringToFloat.enabled': 'true', 'spark.rapids.sql.hashAgg.replaceMode': 'partial'}-[('a', RepeatSeq(Long)), ('b', RepeatSeq(Boolean)), ('c', LongRange(not_null))]][IGNORE_ORDER({'local': True}), APPROXIMATE_FLOAT, ALLOW_NON_GPU(ObjectHashAggregateExec,SortAggregateExec,ShuffleExchangeExec,HashPartitioning,SortExec,SortArray,Alias,Literal,Count,CollectList,CollectSet,GpuToCpuCollectBufferTransition,CpuToGpuCollectBufferTransition,AggregateExpression)]
 and alot of others of that same test.

```
E                   py4j.protocol.Py4JJavaError: An error occurred while calling z:com.nvidia.spark.rapids.ExecutionPlanCaptureCallback.assertContains.
10:05:47  E                   : java.lang.AssertionError: assertion failed: Could not find GpuCollectList in the Spark plan
10:05:47  E                   ObjectHashAggregate(keys=[a#1739656L], functions=[collect_list(b#1739657, 0, 0), collect_set(b#1739657, 0, 0)], output=[a#1739656L, sort_array(collect_list(b), true)#1739667, sort_array(collect_set(b), true)#1739668])
10:05:47  E                   +- GpuColumnarToRow false
10:05:47  E                      +- GpuShuffleCoalesce 2147483647
10:05:47  E                         +- GpuCustomShuffleReader coalesced
10:05:47  E                            +- ShuffleQueryStage 0, Statistics(sizeInBytes=4.3 KiB, rowCount=100, isRuntime=true)
10:05:47  E                               +- GpuColumnarExchange gpuhashpartitioning(a#1739656L, 12), true, [id=#218025]
10:05:47  E                                  +- GpuProject [a#1739656L, b#1739657]
10:05:47  E                                     +- GpuProject [a#1739656L, b#1739657]
10:05:47  E                                        +- GpuRowToColumnar targetsize(2147483647)
10:05:47  E                                           +- *(1) Scan ExistingRDD[a#1739656L,b#1739657,c#1739658L]
10:05:47  E                   
10:05:47  E                   	at scala.Predef$.assert(Predef.scala:223)
10:05:47  E                   	at com.nvidia.spark.rapids.ExecutionPlanCaptureCallback$.assertContains(Plugin.scala:336)
10:05:47  E                   	at com.nvidia.spark.rapids.ExecutionPlanCaptureCallback$.assertContains(Plugin.scala:341)
10:05:47  E                   	at com.nvidia.spark.rapids.ExecutionPlanCaptureCallback.assertContains(Plugin.scala)
10:05:47  E                   	at sun.reflect.GeneratedMethodAccessor472.invoke(Unknown Source)
10:05:47  E                   	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
10:05:47  E                   	at java.lang.reflect.Method.invoke(Method.java:498)
10:05:47  E                   	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
10:05:47  E                   	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:380)
10:05:47  E                   	at py4j.Gateway.invoke(Gateway.java:295)
10:05:47  E                   	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
10:05:47  E                   	at py4j.commands.CallCommand.execute(CallCommand.java:79)
10:05:47  E                   	at py4j.GatewayConnection.run(GatewayConnection.java:251)
10:05:47  E                   	at java.lang.Thread.run(Thread.java:748)
10:05:47  
```

### PR DESCRIPTION
**Pull request by sperlingxx**
Signed-off-by: sperlingxx <lovedreamf@gmail.com>

Fixes #3339

Current PR is to fix the test failure of test_hash_groupby_collect_partial_replace_fallback on Databricks runtime. The root cause of the test failure is about different planning strategies for Aggregation between Spark and Databricks runtime. To support both Spark and Databricks runtime, we need a more expressive configuration for `hashAggReplaceMode`.  In addition, with the extended `TagForReplaceMode` method, we are able to build tests on partial GPU replacement for Aggregate in a more preciser way. 

P.S. It has been manually tested on DB_301.

### CODE REVIEW

**Comment by sperlingxx** [comment]
build

**Comment by sperlingxx** [comment]
build

**Comment by sperlingxx** [comment]
build

File: integration_tests/src/main/python/hash_aggregate_test.py [hunk 1]
```diff
@@ -615,10 +661,9 @@ def test_hash_multiple_mode_query_avg_distincts(data_gen, conf):
 @approximate_float
 @ignore_order
 @incompat
-@pytest.mark.parametrize('data_gen', _init_list_no_nans, ids=idfn)
+@pytest.mark.parametrize('data_gen', _init_list_no_nans[1:2], ids=idfn)
```
Comment (revans2) [inline integration_tests/src/main/python/hash_aggregate_test.py:664]: Using a slice for the data gen is really confusing here. At a minimum we need a comment explaining why we are slicing it. Preferably it is a separate value along with the explanation.

**Review by revans2 — APPROVED** [review]
Just a nit

File: integration_tests/src/main/python/hash_aggregate_test.py (hunk 1 above)
Comment (sperlingxx) [reply integration_tests/src/main/python/hash_aggregate_test.py]: Sorry, its my mistake. I added the slice to facilitate debugging, and I forgot to remove it before submission.

**Comment by sperlingxx** [comment]
build

**Review by revans2 — DISMISSED** [review]
Looks good to me, but I would like to have a few more eyes look at it.

File: integration_tests/src/main/python/hash_aggregate_test.py [hunk 2]
```diff
@@ -617,8 +663,7 @@ def test_hash_multiple_mode_query_avg_distincts(data_gen, conf):
 @incompat
 @pytest.mark.parametrize('data_gen', _init_list_no_nans, ids=idfn)
 @pytest.mark.parametrize('conf', get_params(_confs, params_markers_for_confs), ids=idfn)
-@pytest.mark.parametrize('parameterless', ['true', pytest.param('false', marks=pytest.mark.xfail(
-    condition=not is_before_spark_311(), reason="parameterless count not supported by default in Spark 3.1+"))])
```
Comment (tgravescs) [inline integration_tests/src/main/python/hash_aggregate_test.py:621]: why was this removed?

**Review by abellina — APPROVED** [review]
It seems like a good approach to be able to express the varying nature of these aggregate plans and the prior method was limiting and wrong in some cases. 

I'd like to see these patterns tied to a spark version/flavor (somehow), but that can come in a follow up. @sperlingxx what do you think? Main reason I bring this up is so that it's clearer from the pattern what Spark we are targeting.

File: integration_tests/src/main/python/hash_aggregate_test.py [hunk 3]
```diff
@@ -514,54 +515,99 @@ def spark_fn(spark_session):
...
+            'spark.sql.adaptive.enabled': aqe_enabled}
     # test with single Distinct
     assert_cpu_and_gpu_are_equal_collect_with_capture(
         lambda spark: gen_df(spark, data_gen, length=100)
             .groupby('a')
             .agg(f.sort_array(f.collect_list('b')),
                  f.sort_array(f.collect_set('b')),
-                 f.countDistinct('c'),
-                 f.count('c')),
+                 f.countDistinct('c')),
```
[the remark this answers is not in the document]
Comment (abellina) [reply integration_tests/src/main/python/hash_aggregate_test.py:594]: @sperlingxx  I did have this issue, I am not sure why this test content had to change. That said, I know we are trying to get some tests to pass, as long as we address this either on this PR or a subsequent one linked to this PR.

File: integration_tests/src/main/python/hash_aggregate_test.py (hunk 2 above)
Comment (sperlingxx) [reply integration_tests/src/main/python/hash_aggregate_test.py]: I am sorry. This option shouldn't be removed. I added it back.

**Comment by sperlingxx** [comment]
build

File: integration_tests/src/main/python/hash_aggregate_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sperlingxx) [reply integration_tests/src/main/python/hash_aggregate_test.py:587]: @abellina `f.count` got removed because it looks redundant since we already had non-distinct aggregations: `collect_list` and `collect_set`.

**Comment by sperlingxx** [comment]
> It seems like a good approach to be able to express the varying nature of these aggregate plans and the prior method was limiting and wrong in some cases.
> 
> I'd like to see these patterns tied to a spark version/flavor (somehow), but that can come in a follow up. @sperlingxx what do you think? Main reason I bring this up is so that it's clearer from the pattern what Spark we are targeting.



> It seems like a good approach to be able to express the varying nature of these aggregate plans and the prior method was limiting and wrong in some cases.
> 
> I'd like to see these patterns tied to a spark version/flavor (somehow), but that can come in a follow up. @sperlingxx what do you think? Main reason I bring this up is so that it's clearer from the pattern what Spark we are targeting.

Hi @abellina , I fully agree on the idea about listing common aggregation patterns of spark. Perhaps we can label them as constants? I filed a follow-up issue #3437 for your idea.

File: integration_tests/src/main/python/hash_aggregate_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (abellina) [reply integration_tests/src/main/python/hash_aggregate_test.py:587]: Ok, that's not consistent. There are other tests with that combination and with collect*. I personally would rather have more aggregates computed than less, it is more chance to catch issues with the aggregation buffer/ordinal logic. 

This test also has two assertions, and I recall distinctly having to remove one of them to debug the other. This is different than my previous comment, but it should really be two tests.

This is blocking other things, I am ok doing this as a follow up.

**Review by tgravescs — APPROVED** [review]

**Review by abellina — APPROVED** [review]
